### ISSUE TITLE
Spring REST?

### ISSUE THREAD

**Issue by Rouche** [vojtechhabarta/typescript-generator#253]
Hi, just a fast question. Do you support Spring 5 REST format (@GetMapping for example) and other annotations for parameters/body etc?

I might become contributor to support it if not, but not in immediate time.

Thanks.

**Comment by gordon00** [comment]
+1

**Comment by Rouche** [comment]
For info, Its no longer needed on my side. I implemented it but from our in-house TypeScript generator that was really different and that i knew very well. (Generate POJOs, Services interfaces with their implementations using AxiosResponses)
Sadly i'm not proprietary of the source so i cannot share.

**Comment by vojtechhabarta** [comment]
@Rouche thanks for the update and sharing this example.

**Comment by gordon00** [comment]
@Rouche And thank you very much for your feedback. I would like to see the code you made to generate what you've shown.
thank you

**Comment by Rouche** [comment]
Ill see what i can do. I cannot really split the code i did with the one already made.

**Comment by vojtechhabarta** [comment]
I started working on this. Tricky thing is that one Java method can be mapped to several HTTP methods (GET, POST). What would be preferred solution? Creating separate client method for each HTTP method and appending this HTTP method to original Java method name?

**Comment by Rouche** [comment]
That would be an idea but right now we enforced a convention for defining Services.

- API methods have only one mapping.
- Mapped methods have different name 
```
@GetMapping
getClients()
@PostMapping
saveClient()
```
- No methods overrides
```
@GetMapping
getClientsForSomething()
```

Its simpler and TypeScript stays identical to the Java API. So we can "find all" in the IDE and hit backend and front end at same time.

**Comment by szwoelf** [comment]
We're also trying this. The restirction with `@getMapping` (...) seems very ok.

**Comment by gordon00** [comment]
@vojtechhabarta thank you for working on this. I can't wait to test your work

**Comment by szwoelf** [comment]
Can I help in some way concerning this?

**Comment by vojtechhabarta** [comment]
I started working on refactoring of REST support and implementing support for some basic Spring application. After finishing this work any help with other features will be appreciated.

**Comment by Rouche** [comment]
Sadly i got no news from my old employer to share code.

**Comment by zth** [comment]
I'm very interested in Spring support as well, following this with excitement.

**Comment by vojtechhabarta** [comment]
I finally committed first version of support for Spring REST applications. The refactoring was bigger than expected. Spring support is implemented in separate module `typescript-generator-spring` because I don't want to add few tens of (transitive) dependencies to core module. You have to add this module to build plugin dependencies when you want to use it - see `sample-maven-spring` or `sample-gradle-spring` module for an example.

To generate client class for Spring REST set `generateSpringApplicationClient` parameter to `true` and specify Spring _controller_ classes using for example `classPatterns` parameter. You can also use `scanSpringApplication` parameter and just specify single Spring _application_ class using `classes` parameter. In this case Sprint application is loaded and list of controller classes is obtained from the application.

Any feedback is much appreciated.

**Comment by zth** [comment]
Awesome! I'm trying to test it using the sample config you've set up in the repo, but it doesn't find `typescript-generator-spring` (I can't find anything https://jcenter.bintray.com/cz/habarta/typescript-generator/ here either). Do I need to do something more for it to work?

**Comment by vojtechhabarta** [comment]
@zth actually this commit is not released yet, it is just in `master` branch so you would need to clone the repo and build it yourself (and then use version 2.12-SNAPSHOT).

**Comment by zth** [comment]
@vojtechhabarta ahh of course! I've tried it now and it works great when I configure it to generate a client, but as soon as I turn that off I almost only empty definitions for the different controllers. Any idea what that might be, or am I misunderstanding something? I'd like to get just the types for the various endpoints.

**Comment by vojtechhabarta** [comment]
Do you want to generate `*.d.ts` (declaration file) or `*.ts` (implementation file)?
Let's say we have this example:
``` java
@SpringBootApplication
public class SpringTestApplication {
    @RestController
    public static class GreetingController {
        @RequestMapping("/greeting")
        public Greeting greeting(@RequestParam(value="name") String name) {
        }
    }
    public static class Greeting {
    }
}
```
Do you want to generate just `interface Greeting` or do you want also REST client methods like `greeting(queryParams?: { name?: string; }): RestResponse<Greeting>`?

**Comment by zth** [comment]
I'd say to start with I only want the Greeting interface, but I do see the value of the entire method.

**Comment by vojtechhabarta** [comment]
@zth generating just `Greeting` interface was already possible (before this commit). Just use `classes` or `classPatterns` parameter and specify your JSON classes. What's new is support for Spring REST methods.

**Comment by vojtechhabarta** [comment]
I released [v2.13.489](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.13.489) with initial support for Spring.
